### ISSUE TITLE
1006 (connection closed abnormally [internal]), no reason

### ISSUE THREAD

**Issue by sonicat01** [alpacahq/pylivetrader#127]
Hi, I am using pylivetrader to run my trade algo on a cloud VM. However, after running a while, I keep getting this error message: `1006 (connection closed abnormally [internal]), no reason`. I guess the cloud infra have some TCP timeout mechanism to trigger this websocket closed abnormally.  
I guess the reason I got this error is that the code around here:
```
 def _get_stream(self, context):
        set_context(context)
        asyncio.set_event_loop(asyncio.new_event_loop())
        conn = tradeapi.StreamConn(self._key_id, self._secret, self._base_url)
        channels = ['trade_updates']
``` 
Can you help to identify and fix the issue?  
 
Thank you.

**Comment by umitanuki** [comment]
We have seen frequent disconnection around websocket lately. We are looking into it.

**Comment by shlomikushchi** [comment]
@sonicat01 does this still happen?

**Comment by cunhafh** [comment]
Hi guys!

I'm running into the same problem. I deployed my bot on PythonAnywhere and Heroku and I'm getting disconnected. Any ideas on how to fix it?

Thanks

**Comment by shlomikushchi** [comment]
can you send the stacktrace or logs you get?

**Comment by cunhafh** [comment]
Hi, thanks for coming on the rescue again.

This is the error message Im getting:

2020-04-15T15:46:13.552503+00:00 app[worker.1]: WARNING:root:code = 1006 (connection closed abnormally [internal]), no reason
2020-04-15T15:46:13.576793+00:00 app[worker.1]: WARNING:root:server rejected WebSocket connection: HTTP 502

I deployed it on PythonAnywhere and Heroku, got the same msg on both

**Comment by shlomikushchi** [comment]
this could be a server error (but then I would expect more people would have issues) or execution enivronment issue.
connection errors are tricky because I don't have your environment to debug.. but let's see if I have an idea
I have some questions that may help me understand a bit better what happend

did you try to run this locally as well? if so what happend?
does it happen every time or some of the times? if not always, are you able to run it as expected?
what is the command you used to execute this? (`pylivetrader run ...`)
did you use pipeline-live?
did you just use the example code or did you write something yourself?
do you have a funded alpaca account? 
did you use the paper trading account?

**Comment by shlomikushchi** [comment]
I don't know pythonanywhere that well, but from their website they say that the free instances have limited internet access: 
![image](https://user-images.githubusercontent.com/7924802/79461064-54d6b780-7ffe-11ea-8c0a-8c6892bc3fe1.png)

that may cause that?

**Comment by cunhafh** [comment]
Hi Shlomi,

I have a paid PythonAnywhere account. I have also run it on Heroku, and I get the same error. My orders are getting through but eventually the websocket disconnects.


Let me answer your questions:

did you try to run this locally as well? if so what happend? It also happens when I run it locally

does it happen every time or some of the times? if not always, are you able to run it as expected? I can run it for some hours or minutes at times, my orders are getting executed. But eventually the connection dies.

what is the command you used to execute this? (pylivetrader run …) pylivetrader run -f bb.py —backend-config config.yaml

did you use pipeline-live? Nope

did you just use the example code or did you write something yourself? I’m using my own code

do you have a funded alpaca account? My account is funded

did you use the paper trading account? So, far I’ve only used the paper account


Would you mind to share your set up to run pylivetrader? I’m not hang up to PA nor Heroku.

Thank you

**Comment by sonicat01** [comment]
Hi @shlomikushchi : 
I use google 's VM instance with CentOS 7 environment. I think you can create a VM instance and easily reproduce the issue there. I think your asynico needs to capture this error and reconnect the WebSocket. It should resolve the issue.

Here are more logs that may help you to investigate the issue:
My code only use pylivetrader, no pipeline-live.

```
WARNING:root:code = 1006 (connection closed abnormally [internal]), no reason
WARNING:root:code = 1006 (connection closed abnormally [internal]), no reason
WARNING:root:server rejected WebSocket connection: HTTP 502
ERROR:asyncio:Future exception was never retrieved
future: <Future finished exception=ConnectionClosedError('code = 1006 (connection closed abnormally [internal]), no reason',)>
websockets.exceptions.ConnectionClosedError: code = 1006 (connection closed abnormally [internal]), no reason
WARNING:root:code = 1006 (connection closed abnormally [internal]), no reason
```
> this could be a server error (but then I would expect more people would have issues) or execution enivronment issue.
> connection errors are tricky because I don't have your environment to debug.. but let's see if I have an idea
> I have some questions that may help me understand a bit better what happend
> 
> did you try to run this locally as well? if so what happend?
> does it happen every time or some of the times? if not always, are you able to run it as expected?
> what is the command you used to execute this? (`pylivetrader run ...`)
> did you use pipeline-live?
> did you just use the example code or did you write something yourself?
> do you have a funded alpaca account?
> did you use the paper trading account?

**Comment by shlomikushchi** [comment]
so if I understand correctly we are dealing with a websocket that after a period of time (maybe minutes, maybe hours) disconnects and doesn't reconnect.
I will try to reproduce it locally and will let you know what I find.

**Comment by shlomikushchi** [comment]
@cunhafh I run it locally. not in a cloud service

**Comment by shlomikushchi** [comment]
one more question? does it happen during trading hours? or after market close?

**Comment by cunhafh** [comment]
Shlomi,

Your description of the problem is precisely what I’m experiencing.

I’m not sure if it happens when the market is closed. Today I restarted my algo 2 times (during market hours).

**Comment by shlomikushchi** [comment]
so I am able to get something similar locally, but when it happens locally it doesn't stop. it continues.
look:
![image](https://user-images.githubusercontent.com/7924802/79595699-fdad1180-80e7-11ea-9059-f49553a2ab44.png)
is that what happens in your scenario? or does it stop working once that happens?

**Comment by cunhafh** [comment]
Hi

Yes, this is the error Im getting. Once it happens I kill the bash and
restart the bot, it doesnt get back to normal by itself

**Comment by shlomikushchi** [comment]
so maybe it's not the same issue because here it continues.
do you use the latest version of pylivetrader?

**Comment by shlomikushchi** [comment]
let's make sure we use the same code

**Comment by shlomikushchi** [comment]
latest version I mean the one on github.
install it like this: 
`pip install git+https://github.com/alpacahq/pylivetrader/`

**Comment by shlomikushchi** [comment]
I have a patch for this issue.
before I merge it, could you check that it works for your scenario?
you need to manually install these 2 packages:
```
pip install -U git+https://github.com/alpacahq/pylivetrader@fix_streamcon_connection_loss
pip install -U git+https://github.com/shlomikushchi/alpaca-trade-api-python@stream2_exception_handling
```

please let me know if that works for you too

**Comment by cunhafh** [comment]
Hi, thanks for your reply.

I’ve updated the packages as per your instructions and am currently testing it. I’ll let you know the results tomorrow.

Regards

**Comment by cunhafh** [comment]
Hi Shlomi,

I still get the disconnect error but now the execution seems to be resuming. 

Thanks for your help.

I’m attaching the last log for your reference.

(venv) 18:52 ~ $ pylivetrader run -f bb.py --backend-config config.yaml
[2020-04-19 18:53:05.622426] INFO: Algorithm: livetrader start running with backend = alpaca data-frequency = minute
Initializing...
WARNING:root:code = 1006 (connection closed abnormally [internal]), no reason
WARNING:root:code = 1006 (connection closed abnormally [internal]), no reason
WARNING:root:code = 1006 (connection closed abnormally [internal]), no reason
WARNING:root:server rejected WebSocket connection: HTTP 502
WARNING:root:code = 1006 (connection closed abnormally [internal]), no reason
ERROR:asyncio:Task exception was never retrieved
future: <Task finished coro=<StreamConn._ensure_ws() done, defined at /home/cunha/venv/lib/python3.6/site-packages/alpaca_trade_api/stream2.py:88> exception=ConnectionRefusedError(111, "Connect call failed ('35.194.67.18', 443)")>
Traceback (most recent call last):
  File "/home/cunha/venv/lib/python3.6/site-packages/alpaca_trade_api/stream2.py", line 94, in _ensure_ws
    await self._connect()
  File "/home/cunha/venv/lib/python3.6/site-packages/alpaca_trade_api/stream2.py", line 35, in _connect
    ws = await websockets.connect(self._endpoint)
  File "/home/cunha/venv/lib/python3.6/site-packages/websockets/client.py", line 535, in __await_impl__
    transport, protocol = await self._create_connection()
  File "/usr/lib/python3.6/asyncio/base_events.py", line 794, in create_connection
    raise exceptions[0]
  File "/usr/lib/python3.6/asyncio/base_events.py", line 781, in create_connection
    yield from self.sock_connect(sock, address)
  File "/usr/lib/python3.6/asyncio/selector_events.py", line 439, in sock_connect
    return (yield from fut)
  File "/usr/lib/python3.6/asyncio/selector_events.py", line 469, in _sock_connect_cb
    raise OSError(err, 'Connect call failed %s' % (address,))
ConnectionRefusedError: [Errno 111] Connect call failed ('35.194.67.18', 443)
[2020-04-20 13:31:00.305016] ERROR: Executor: Parser must be a string or character stream, not int
Traceback (most recent call last):
  File "/home/cunha/venv/lib/python3.6/site-packages/pylivetrader/executor/executor.py", line 67, in wrapper
    func(*args, **kwargs)
  File "/home/cunha/venv/lib/python3.6/site-packages/pylivetrader/executor/executor.py", line 88, in every_bar
    handle_data(algo, current_data, dt_to_use)
  File "/home/cunha/venv/lib/python3.6/site-packages/pylivetrader/misc/events.py", line 218, in handle_data
    dt,
  File "/home/cunha/venv/lib/python3.6/site-packages/pylivetrader/misc/events.py", line 237, in handle_data
    self.callback(context, data)
  File "bb.py", line 32, in bollinger_bands
    current_price = data.current(context.stock, 'price')
  File "/home/cunha/venv/lib/python3.6/site-packages/pylivetrader/data/bardata.py", line 85, in current
    results = parallelize(fetch)(fetch_args)
  File "/home/cunha/venv/lib/python3.6/site-packages/pylivetrader/misc/parallel_utils.py", line 39, in wrapper
    task_result = task.result()
  File "/usr/lib/python3.6/concurrent/futures/_base.py", line 425, in result
    return self.__get_result()
  File "/usr/lib/python3.6/concurrent/futures/_base.py", line 384, in __get_result
    raise self._exception
  File "/usr/lib/python3.6/concurrent/futures/thread.py", line 56, in run
    result = self.fn(*self.args, **self.kwargs)
  File "/home/cunha/venv/lib/python3.6/site-packages/pylivetrader/data/bardata.py", line 73, in fetch
    self.data_frequency
  File "/home/cunha/venv/lib/python3.6/site-packages/pylivetrader/data/data_portal.py", line 56, in get_spot_value
    assets, field, dt, data_frequency, self.quantopian_compatible
  File "/home/cunha/venv/lib/python3.6/site-packages/pylivetrader/backend/alpaca.py", line 445, in get_spot_value
    results = self._get_spot_bars(symbols, field)
  File "/home/cunha/venv/lib/python3.6/site-packages/pylivetrader/backend/alpaca.py", line 466, in _get_spot_bars
    symbol_bars = self._symbol_bars(symbols, 'minute', limit=1)
  File "/home/cunha/venv/lib/python3.6/site-packages/pylivetrader/backend/alpaca.py", line 620, in _symbol_bars
    return parallelize(fetch)(symbols)
  File "/home/cunha/venv/lib/python3.6/site-packages/pylivetrader/misc/parallel_utils.py", line 39, in wrapper
    task_result = task.result()
  File "/usr/lib/python3.6/concurrent/futures/_base.py", line 425, in result
    return self.__get_result()
  File "/usr/lib/python3.6/concurrent/futures/_base.py", line 384, in __get_result
    raise self._exception
  File "/usr/lib/python3.6/concurrent/futures/thread.py", line 56, in run
    result = self.fn(*self.args, **self.kwargs)
  File "/home/cunha/venv/lib/python3.6/site-packages/pylivetrader/backend/alpaca.py", line 82, in wrapper
    return func(*args, **kwargs)
  File "/home/cunha/venv/lib/python3.6/site-packages/pylivetrader/backend/alpaca.py", line 583, in fetch
    int(to.timestamp()) * 1000
  File "/home/cunha/venv/lib/python3.6/site-packages/alpaca_trade_api/polygon/rest.py", line 145, in historic_agg_v2
    _from=dateutil.parser.parse(_from).date().isoformat(),
  File "/home/cunha/venv/lib/python3.6/site-packages/dateutil/parser/_parser.py", line 1374, in parse
    return DEFAULTPARSER.parse(timestr, **kwargs)
  File "/home/cunha/venv/lib/python3.6/site-packages/dateutil/parser/_parser.py", line 646, in parse
    res, skipped_tokens = self._parse(timestr, **kwargs)
  File "/home/cunha/venv/lib/python3.6/site-packages/dateutil/parser/_parser.py", line 725, in _parse
    l = _timelex.split(timestr)         # Splits the timestr into tokens
  File "/home/cunha/venv/lib/python3.6/site-packages/dateutil/parser/_parser.py", line 207, in split
    return list(cls(s))
  File "/home/cunha/venv/lib/python3.6/site-packages/dateutil/parser/_parser.py", line 76, in __init__
    '{itype}'.format(itype=instream.__class__.__name__))
TypeError: Parser must be a string or character stream, not int
[2020-04-20 13:31:00.512557] WARNING: Executor: Continuing execution

**Comment by shlomikushchi** [comment]
disconnection is something that could happen of course. if it reconnects it means we're handling it.
I will leave this issue opened for a few more days. let me know if it works or if the solution is not resilient enough

**Comment by cunhafh** [comment]
Hi Shlomi,

I’ve experienced a different disconnect error this morning. I’m attaching the log for your reference.

In order to eliminate variables, would you be able to share the quantopian script you’re running so I can ran it from my end?

Thanks

2020-04-20 18:03:46 - Task preparing to start
Apr 20 18:07:11 [2020-04-20 18:07:11.469071] INFO: Algorithm: livetrader start running with backend = alpaca data-frequency = minute
Apr 20 18:08:00 Initializing...
Apr 20 18:08:00 [2020-04-20 18:08:00.113853] ERROR: Executor: Parser must be a string or character stream, not int
Apr 20 18:08:00 Traceback (most recent call last):
Apr 20 18:08:00   File "/home/cunha/venv/lib/python3.6/site-packages/pylivetrader/executor/executor.py", line 67, in wrapper
Apr 20 18:08:00     func(*args, **kwargs)
Apr 20 18:08:00   File "/home/cunha/venv/lib/python3.6/site-packages/pylivetrader/executor/executor.py", line 88, in every_bar
Apr 20 18:08:00     handle_data(algo, current_data, dt_to_use)
Apr 20 18:08:00   File "/home/cunha/venv/lib/python3.6/site-packages/pylivetrader/misc/events.py", line 218, in handle_data
Apr 20 18:08:00     dt,
Apr 20 18:08:00   File "/home/cunha/venv/lib/python3.6/site-packages/pylivetrader/misc/events.py", line 237, in handle_data
Apr 20 18:08:00     self.callback(context, data)
Apr 20 18:08:00   File "bb.py", line 32, in bollinger_bands
Apr 20 18:08:00     current_price = data.current(context.stock, 'price')
Apr 20 18:08:00   File "/home/cunha/venv/lib/python3.6/site-packages/pylivetrader/data/bardata.py", line 85, in current
Apr 20 18:08:00     results = parallelize(fetch)(fetch_args)
Apr 20 18:08:00   File "/home/cunha/venv/lib/python3.6/site-packages/pylivetrader/misc/parallel_utils.py", line 39, in wrapper
Apr 20 18:08:00     task_result = task.result()
Apr 20 18:08:00   File "/usr/lib/python3.6/concurrent/futures/_base.py", line 425, in result
Apr 20 18:08:00     return self.__get_result()
Apr 20 18:08:00   File "/usr/lib/python3.6/concurrent/futures/_base.py", line 384, in __get_result
Apr 20 18:08:00     raise self._exception
Apr 20 18:08:00   File "/usr/lib/python3.6/concurrent/futures/thread.py", line 56, in run
Apr 20 18:08:00     result = self.fn(*self.args, **self.kwargs)
Apr 20 18:08:00   File "/home/cunha/venv/lib/python3.6/site-packages/pylivetrader/data/bardata.py", line 73, in fetch
Apr 20 18:08:00     self.data_frequency
Apr 20 18:08:00   File "/home/cunha/venv/lib/python3.6/site-packages/pylivetrader/data/data_portal.py", line 56, in get_spot_value
Apr 20 18:08:00     assets, field, dt, data_frequency, self.quantopian_compatible
Apr 20 18:08:00   File "/home/cunha/venv/lib/python3.6/site-packages/pylivetrader/backend/alpaca.py", line 445, in get_spot_value
Apr 20 18:08:00     results = self._get_spot_bars(symbols, field)
Apr 20 18:08:00   File "/home/cunha/venv/lib/python3.6/site-packages/pylivetrader/backend/alpaca.py", line 466, in _get_spot_bars
Apr 20 18:08:00     symbol_bars = self._symbol_bars(symbols, 'minute', limit=1)
Apr 20 18:08:00   File "/home/cunha/venv/lib/python3.6/site-packages/pylivetrader/backend/alpaca.py", line 620, in _symbol_bars
Apr 20 18:08:00     return parallelize(fetch)(symbols)
Apr 20 18:08:00   File "/home/cunha/venv/lib/python3.6/site-packages/pylivetrader/misc/parallel_utils.py", line 39, in wrapper
Apr 20 18:08:00     task_result = task.result()
Apr 20 18:08:00   File "/usr/lib/python3.6/concurrent/futures/_base.py", line 425, in result
Apr 20 18:08:00     return self.__get_result()
Apr 20 18:08:00   File "/usr/lib/python3.6/concurrent/futures/_base.py", line 384, in __get_result
Apr 20 18:08:00     raise self._exception
Apr 20 18:08:00   File "/usr/lib/python3.6/concurrent/futures/thread.py", line 56, in run
Apr 20 18:08:00     result = self.fn(*self.args, **self.kwargs)
Apr 20 18:08:00   File "/home/cunha/venv/lib/python3.6/site-packages/pylivetrader/backend/alpaca.py", line 82, in wrapper
Apr 20 18:08:00     return func(*args, **kwargs)
Apr 20 18:08:00   File "/home/cunha/venv/lib/python3.6/site-packages/pylivetrader/backend/alpaca.py", line 583, in fetch
Apr 20 18:08:00     int(to.timestamp()) * 1000
Apr 20 18:08:00   File "/home/cunha/venv/lib/python3.6/site-packages/alpaca_trade_api/polygon/rest.py", line 145, in historic_agg_v2
Apr 20 18:08:00     _from=dateutil.parser.parse(_from).date().isoformat(),
Apr 20 18:08:00   File "/home/cunha/venv/lib/python3.6/site-packages/dateutil/parser/_parser.py", line 1374, in parse
Apr 20 18:08:00     return DEFAULTPARSER.parse(timestr, **kwargs)
Apr 20 18:08:00   File "/home/cunha/venv/lib/python3.6/site-packages/dateutil/parser/_parser.py", line 646, in parse
Apr 20 18:08:00     res, skipped_tokens = self._parse(timestr, **kwargs)
Apr 20 18:08:00   File "/home/cunha/venv/lib/python3.6/site-packages/dateutil/parser/_parser.py", line 725, in _parse
Apr 20 18:08:00     l = _timelex.split(timestr)         # Splits the timestr into tokens
Apr 20 18:08:00   File "/home/cunha/venv/lib/python3.6/site-packages/dateutil/parser/_parser.py", line 207, in split
Apr 20 18:08:00     return list(cls(s))
Apr 20 18:08:00   File "/home/cunha/venv/lib/python3.6/site-packages/dateutil/parser/_parser.py", line 76, in __init__
Apr 20 18:08:00     '{itype}'.format(itype=instream.__class__.__name__))
Apr 20 18:08:00 TypeError: Parser must be a string or character stream, not int
Apr 20 18:08:00 [2020-04-20 18:08:00.160456] WARNING: Executor: Continuing execution
Apr 21 01:08:43 WARNING:root:code = 1006 (connection closed abnormally [internal]), no reason
Apr 21 01:10:36 WARNING:root:code = 1006 (connection closed abnormally [internal]), no reason
Apr 21 01:10:36 ERROR:asyncio:Task exception was never retrieved
Apr 21 01:10:36 future: <Task finished coro=<StreamConn._ensure_ws() done, defined at /home/cunha/venv/lib/python3.6/site-packages/alpaca_trade_api/stream2.py:88> exception=ConnectionRefusedError(111, "Connect call failed ('35.194.67.18', 443)")>
Apr 21 01:10:36 Traceback (most recent call last):
Apr 21 01:10:36   File "/home/cunha/venv/lib/python3.6/site-packages/alpaca_trade_api/stream2.py", line 94, in _ensure_ws
Apr 21 01:10:36     await self._connect()
Apr 21 01:10:36   File "/home/cunha/venv/lib/python3.6/site-packages/alpaca_trade_api/stream2.py", line 35, in _connect
Apr 21 01:10:36     ws = await websockets.connect(self._endpoint)
Apr 21 01:10:36   File "/home/cunha/venv/lib/python3.6/site-packages/websockets/client.py", line 535, in __await_impl__
Apr 21 01:10:36     transport, protocol = await self._create_connection()
Apr 21 01:10:36   File "/usr/lib/python3.6/asyncio/base_events.py", line 794, in create_connection
Apr 21 01:10:36     raise exceptions[0]
Apr 21 01:10:36   File "/usr/lib/python3.6/asyncio/base_events.py", line 781, in create_connection
Apr 21 01:10:36     yield from self.sock_connect(sock, address)
Apr 21 01:10:36   File "/usr/lib/python3.6/asyncio/selector_events.py", line 439, in sock_connect
Apr 21 01:10:36     return (yield from fut)
Apr 21 01:10:36   File "/usr/lib/python3.6/asyncio/selector_events.py", line 469, in _sock_connect_cb
Apr 21 01:10:36     raise OSError(err, 'Connect call failed %s' % (address,))
Apr 21 01:10:36 ConnectionRefusedError: [Errno 111] Connect call failed ('35.194.67.18', 443)

**Comment by shlomikushchi** [comment]
this is a different issue so let's open a new issue so we don't get people mixed up.
delete your last comment and open a new issue.
I will give you my insights there 
(basically, the server disconnects you at night)

**Comment by lcwarrin** [comment]
Hi, I have been running in to a similar issue:
I have set up everything to run based on the api documentation but I keep receiving  the following errors:WARNING:root:code = 1006 (connection closed abnormally [internal]), no reasonWARNING:root:server rejected WebSocket connection: HTTP 502WARNING:root:code = 1006 (connection closed abnormally [internal]), no reasonConnectionRefusedError: [Errno 111] Connect call failed ('35.221.23.121', 443)or .HTTPError: 500 Server Error

When I tried using the solution mentioned above, which is  to install the package pip install -U git+https://github.com/alpacahq/pylve ivetrader@fix_streamcon_connection_loss ,I got the following error:
![image3](https://user-images.githubusercontent.com/19857738/80420413-59258f00-88a0-11ea-990d-2c889dd22ce7.png)


This is the error I received intially:

![alpacaImg (3)](https://user-images.githubusercontent.com/19857738/80419313-8709d400-889e-11ea-8b7e-212f8a7c14ce.png)

**Comment by shlomikushchi** [comment]
Hi. so this branch was merged to the master branch.
install `pip install git+https://github.com/alpacahq/pylivetrader`
and let's see if that is still an issue

**Comment by lcwarrin** [comment]
I installed by entering pip install git+https://github.com/alpacahq/pylivetrader  but still receiving the 1006 error:
![Screenshot (1)](https://user-images.githubusercontent.com/19857738/80543950-94de5880-8975-11ea-9571-ff5885196a7f.png)

I am also using pythonanywhere.

**Comment by shlomikushchi** [comment]
so this is a known issue (which is basically the source of all messages in this page) and I explained it in the related issue we've opened but not yet here. so this is the explination:
we experieince websocket disconnections and this is something we are now working on.
BUT - it should not influence the pylivetrader workflow.
why?
there are 2 ways to get our code called in the alpaca ecosystem:
1. event driven - e.g a new quote was received
2. clock driven - e.g do something every minute

pylivetrader is clock driven
requesting data from the server is done via REST calls.
open/close positions is done as well - via the REST API
the websocket is used to update internal objects (e.g order is filled, update the pylivetrader instance it was filled)
so.. yes it is an issue and we try to solve.
but no, it should not cause issues for pylivetrader users

I hope that gives a good explination for this issue, and I will update you when we solve it.
*if you get other messages or stacktraces please let me know

**Comment by shlomikushchi** [comment]
closing this issue for lack of activity. let me know if it still happens

**Comment by nelfata** [comment]
Hello,
I am still experiencing connection issues quite regularly (every 10 minutes or so):
2020-08-11 11:11:25:   stream2.py:     _consume_msg() 82:WARNING: root : code = 1006 (connection closed abnormally [internal]), no reason

I have what I think the latest code running with Python 3.7.3

I have modified the stream2.py _StreamConn close():
```    async def close(self):
        logging.info("close")
        if self._consume_task:
            logging.info("canceling")
            self._consume_task.cancel()
            logging.info("_consume_task cancelled")
        if self._ws:
            logging.info("closing ws")
            if False:
                await self._ws.close()
            else:
                self._ws.close()
                await self._ws.wait_closed()
            logging.info("closing done")
            self._ws = None
```
The original code uses: await self._ws.close()
But based on the documentation of Asyncio this is not correct (https://docs.python.org/3/library/asyncio-stream.html)
 If I use the modified code I get the following:

```
2020-08-11 11:11:25:   stream2.py:     _consume_msg() 82:WARNING: root : code = 1006 (connection closed abnormally [internal]), no reason
2020-08-11 11:11:25:   stream2.py:            close()133:   INFO: root : close
2020-08-11 11:11:25:   stream2.py:            close()135:   INFO: root : canceling
2020-08-11 11:11:25:   stream2.py:            close()137:   INFO: root : _consume_task cancelled
2020-08-11 11:11:25:   stream2.py:            close()139:   INFO: root : closing ws
/usr/local/lib/python3.7/site-packages/alpaca_trade_api/stream2.py:143: RuntimeWarning: coroutine 'WebSocketCommonProtocol.close' was never awaited
  self._ws.close()
RuntimeWarning: Enable tracemalloc to get the object allocation traceback
2020-08-11 11:11:25:   stream2.py:            close()145:   INFO: root : closing done
2020-08-11 11:11:25:   stream2.py:     _consume_msg() 84:   INFO: root : connection error, closed
2020-08-11 11:11:25:   stream2.py:       _ensure_ws() 88:   INFO: root : _ensure_ws
2020-08-11 11:11:25:   stream2.py:       _ensure_ws() 92:   INFO: root : connect
2020-08-11 11:11:25:   stream2.py:       _ensure_ws() 94:   INFO: root : connecting, retry: 0
2020-08-11 11:11:30:   stream2.py:         _connect() 63:   INFO: root : connected to: wss://data.alpaca.markets/stream
2020-08-11 11:11:30:base_events.py:default_exception_handler()1608:  ERROR: asyncio: Task exception was never retrieved
future: <Task finished coro=<_StreamConn._consume_msg() done, defined at /usr/local/lib/python3.7/site-packages/alpaca_trade_api/stream2.py:70> exception=CancelledError()>
concurrent.futures._base.CancelledError
2020-08-11 11:11:30:   stream2.py:       _ensure_ws() 88:   INFO: root : _ensure_ws
2020-08-11 11:11:30:   stream2.py:       _ensure_ws() 90:   INFO: root : _ws not none
2020-08-11 11:11:30:base_events.py:default_exception_handler()1608:  ERROR: asyncio: Task exception was never retrieved
future: <Task finished coro=<_StreamConn._ensure_ws() done, defined at /usr/local/lib/python3.7/site-packages/alpaca_trade_api/stream2.py:87> exception=TypeError('Object of type set is not JSON serializable')>
Traceback (most recent call last):
  File "/usr/local/lib/python3.7/site-packages/alpaca_trade_api/stream2.py", line 98, in _ensure_ws
    await self.subscribe(self._streams)
  File "/usr/local/lib/python3.7/site-packages/alpaca_trade_api/stream2.py", line 117, in subscribe
    'streams': channels,
  File "/usr/local/lib/python3.7/json/__init__.py", line 231, in dumps
    return _default_encoder.encode(obj)
  File "/usr/local/lib/python3.7/json/encoder.py", line 199, in encode
    chunks = self.iterencode(o, _one_shot=True)
  File "/usr/local/lib/python3.7/json/encoder.py", line 257, in iterencode
    return _iterencode(o, 0)
  File "/usr/local/lib/python3.7/json/encoder.py", line 179, in default
    raise TypeError(f'Object of type {o.__class__.__name__} '
TypeError: Object of type set is not JSON serializable
2020-08-11 11:21:25:   stream2.py:     _consume_msg() 82:WARNING: root : code = 1006 (connection closed abnormally [internal]), no reason
2020-08-11 11:21:25:   stream2.py:            close()133:   INFO: root : close
2020-08-11 11:21:25:   stream2.py:            close()135:   INFO: root : canceling
2020-08-11 11:21:25:   stream2.py:            close()137:   INFO: root : _consume_task cancelled
2020-08-11 11:21:25:   stream2.py:            close()139:   INFO: root : closing ws
2020-08-11 11:21:25:   stream2.py:            close()145:   INFO: root : closing done
2020-08-11 11:21:25:   stream2.py:     _consume_msg() 84:   INFO: root : connection error, closed
2020-08-11 11:21:25:   stream2.py:       _ensure_ws() 88:   INFO: root : _ensure_ws
2020-08-11 11:21:25:   stream2.py:       _ensure_ws() 92:   INFO: root : connect
2020-08-11 11:21:25:   stream2.py:       _ensure_ws() 94:   INFO: root : connecting, retry: 0
```
Please advise, I can send the whole file if need be.
Thanks.

**Comment by shlomikushchi** [comment]
Hi,
you get this warning at the beginning of the stacktrace: `RuntimeWarning: coroutine 'WebSocketCommonProtocol.close' was never awaited`
and it's because of your change.
not sure what you are trying to do.. do remember that ws is not a stream (it contains a stream)

**Comment by nelfata** [comment]
Thanks for your reply. 
I was trying to close the connection based on what the Python async library mentioned in the link I provided.
The reason was that the code was getting stuck in: await self._ws.close()
```
if self._ws:
   await self._ws.close()
```
The connection once it fails never recovers, the on_bar_min() would never get called, but the requests to for other API calls would work just fine.

I think that there is still a bug somewhere preventing from reestablishing the connection.

**Comment by nelfata** [comment]
This is becoming a big issue lately most of this week I get the connection error 1006 every few minutes. I see in the stream2.py that there is a retry to reconnect, but it does not work. Also I am registering with more than 100 symbols. I doubt that this could have any effect.
Any advice would be appreciated. Thank you.

**Comment by shlomikushchi** [comment]
Hi, I do not see these issues.
if you can, provide a code example I could execute and try to get the same errors you get.

**Comment by nelfata** [comment]
Thanks for your reply. It will not be easy but I will try to prepare some code for you. The code is based on the examples with asyncio.
Do you think Python 3.8 would be better dealing with this?

**Comment by shlomikushchi** [comment]
the recommended version is 3.6

**Comment by nelfata** [comment]
I have made a simpler code example. Would it be possible to send you a zip file in private?

**Comment by nelfata** [comment]
I can add another piece of information. Just run a test which establishes a connection with Alpaca for example using stream2.py as per the examples, then disconnect your computer from the network, this will create the websocket 1006 warning message in stream2.py, which is followed by the closing of the socket. The closing will hang indefinitely.
I tried the code on Python 3.6, 3.7 and 3.8 and all of them exhibit the same behavior.
Connecting with close_timeout=3 does not make any difference.

```
    async def _consume_msg(self):
        ws = self._ws
        try:
            while True:
                r = await ws.recv()
                if isinstance(r, bytes):
                    r = r.decode('utf-8')
                msg = json.loads(r)
                stream = msg.get('stream')
                if stream is not None:
                    await self._dispatch(stream, msg)
        except websockets.WebSocketException as wse:
            logging.warn(wse)
            await self.close()                         <<<<<<<<<<<< CODE HANGS HERE
            asyncio.ensure_future(self._ensure_ws())
```

**Comment by nelfata** [comment]
Some more info... I have not done extensive tests but I have been running my application on Alpine Linux from within a Docker container. Running my application outside Docker seems to have much less 1006 errors.
Is this possible?

**Comment by nelfata** [comment]
I sent you a zip file on Slack. 
My point is that irrelevant of the network error I have been having (related to 1006) the internal functionality in Stream2.py should be able to handle it and not block waiting for close(). Once this occurs the minute bars stop updating, but other accesses are fine (for example getting positions etc...), in the example I sent you, I am using Cmd2 library which allows me to send these requests (positions...) to the Alpaca API in parallel.
I only wanted to demonstrate that by disconnecting the network the hang occurs, because the other error which causes the minute bars to stop occurs every about 20 minutes. This has been happening lately quite often. I don't think it is my network connection, as everything else is running fine. For some reason the websocket is not happy.

**Comment by dc500** [comment]
> I don't know pythonanywhere that well, but from their website they say that the free instances have limited internet access:
> ![image](https://user-images.githubusercontent.com/7924802/79461064-54d6b780-7ffe-11ea-8c0a-8c6892bc3fe1.png)
> 
> that may cause that?

It's September and I'm having the same issue. I'm running it locally.

**Comment by shlomikushchi** [comment]
Hi, what exactly do you get?

**Comment by jhills5821** [comment]
Hi, 

I get this error also. I am attempting receive trade updates but once I see the 1006 I will no longer receive any messages. My observations have not been able to completely blame the client side connection because I also connect to polygon's bar websocket and that has continued to return messages after the 1006, however that behavior doesn't necessarily prove causation. 

I would be happy if I could just trigger a retry/new connection from the 1006 but I don't know how to grab it. 

Any thoughts or help would be appreciated.

**Comment by shlomikushchi** [comment]
@jhills5821 try to isolate your connection. do only one type (say trade_updates or bar data) and see what you get when that happens

**Comment by bwees** [comment]
I am having this issue as well. This specifically started when I started routing traffic through @shlomikushchi alpaca proxy. Is this due to rate limiting?

**Comment by shlomikushchi** [comment]
could be. check the logs of the proxy-agent

**Comment by bwees** [comment]
I checked and there is zero output for the proxy. Other than the ascii art and the two statements saying the version and saying that it is using the alpaca web socket, there is nothing.

Brandon Wees

**Comment by shlomikushchi** [comment]
you could try running the proxy-agent locally and use a debugger to see what's wrong. or give me a code sample to execute locally
